File: app/src/main/java/it/corelab/airbooks/section/login/fragment/RecoverPassword_Fragment.java
package it.corelab.airbooks.section.login.fragment;

import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import java.util.Locale;

import cn.pedant.SweetAlert.SweetAlertDialog;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import it.corelab.airbooks.R;
import it.corelab.airbooks.data.model.PostRecoverResponse;
import it.corelab.airbooks.data.model.remote.APIService;
import it.corelab.airbooks.data.model.remote.ApiUtils;
import it.corelab.airbooks.section.login.activity.Login;
import it.corelab.airbooks.section.login.fragment.LoginFragment;

import static android.content.ContentValues.TAG;


public class RecoverPassword_Fragment extends Fragment implements View.OnClickListener {

    protected View view;
    private FragmentManager fragmentManager;
    private TextInputEditText email;
    private Button recover;
    public static String emailAddress;
    private APIService mAPIService;

    public RecoverPassword_Fragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.recover_password, container, false);
        mAPIService = ApiUtils.getAPIService();
        initViews();
        setListeners();
        emailAddress = email.getText().toString();
        return view;
    }

    public void initViews() {
        fragmentManager = getActivity().getSupportFragmentManager();
        email = view.findViewById(R.id.edit_text);
        recover = view.findViewById(R.id.recover_btn);
    }

    public void setListeners() {
        Login.Companion.getLeftArrow().setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                fragmentManager
                        .beginTransaction()
                        .setCustomAnimations(R.anim.right_enter_animation, R.anim.left_exit_animation)
                        .replace(R.id.FRAME_CONTAINER_LOGIN_ACTIVITY, new LoginFragment(), "Login_fragment")
                        .commit();
                setOffLeftArrow();
            }
        });
        recover.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.recover_btn:

                verifyCredential();

                if (isValidCredential()) {

                    Log.i("Credentials", "OK");
                    sendPost(email.getText().toString(), "http://airbooks.altervista.org/API/v2/recover.php?email=" + email.getText().toString(), Locale.getDefault().getLanguage(), "android");

                }
        }
    }

    public void setOffLeftArrow() {
        Login.Companion.getLeftArrow().setEnabled(false);
        Login.Companion.getLeftArrow().setVisibility(View.INVISIBLE);
    }

    boolean isEmailValid(CharSequence email) {
        return android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }

    public String getString(TextInputEditText textInputEditText) {
        String stringa = textInputEditText.getText().toString();
        return stringa;
    }

    public void verifyCredential() {
        if (!isEmailValid(getString(email))) {
            email.setError("you've to insert an email");
        }
    }

    public boolean isValidCredential() {
        return isEmailValid(getString(email));
    }

    public void createDialog() {
        new SweetAlertDialog(getActivity(), SweetAlertDialog.WARNING_TYPE)
                .setTitleText("Recupero password")
                .setContentText("È stata inviata una mail per il recupero password al tuo indirizzo di posta elettronica")
                .setConfirmText("Ho capito")
                .setConfirmClickListener(new SweetAlertDialog.OnSweetClickListener() {
                    @Override
                    public void onClick(SweetAlertDialog sweetAlertDialog) {

                        fragmentManager
                                .beginTransaction()
                                .setCustomAnimations(R.anim.right_enter_animation, R.anim.left_exit_animation)
                                .replace(R.id.FRAME_CONTAINER_LOGIN_ACTIVITY, new LoginFragment(), "Login_fragment")
                                .commit();
                        setOffLeftArrow();

                        sweetAlertDialog.dismiss();
                    }
                })
                .show();
    }

    public void sendPost(String emailText, String url, String lang, String os) {


        final SweetAlertDialog pDialog = new SweetAlertDialog(getActivity(), SweetAlertDialog.PROGRESS_TYPE);
        pDialog.getProgressHelper().setBarColor(Color.parseColor("#4990e2"));
        pDialog.setTitleText("Loading");
        pDialog.setCancelable(false);
        pDialog.show();

        mAPIService.savePost(emailText,url, lang, os).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(new Observer<PostRecoverResponse>() {

            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onNext(PostRecoverResponse postRecoverResponse) {
                pDialog.dismiss();
                    if (postRecoverResponse.getError() != null){
                        Log.i(TAG, "post submitted to API. " + postRecoverResponse.getError().getCode().toString());
                        showErrorDialog();
                    }else {
                        createDialog();
                    }
            }

            @Override
            public void onError(Throwable e) {

            }

            @Override
            public void onComplete() {

            }
        });
    }

    public void showErrorDialog(){
        new SweetAlertDialog(getActivity(), SweetAlertDialog.ERROR_TYPE)
                .setTitleText("Errore")
                .setContentText("Mail non esistente")
                .show();
    }

}

File: app/src/main/java/it/corelab/studios/airbooks/view/fragment/login/RecoverPasswordFragment.kt
package it.corelab.studios.airbooks.view.fragment.login

import android.graphics.Color
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button

import java.util.Locale

import cn.pedant.SweetAlert.SweetAlertDialog
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import it.corelab.studios.airbooks.R
import it.corelab.studios.airbooks.model.data.LOGIN.RECOVERPASSWORD.PostRecoverResponse
import it.corelab.studios.airbooks.model.API.remote.APIService
import it.corelab.studios.airbooks.model.API.remote.ApiUtils
import android.content.ContentValues.TAG
import android.widget.EditText
import it.corelab.studios.airbooks.model.interfaces.login.recover.RecoverPasswordController
import it.corelab.studios.airbooks.model.interfaces.login.recover.TransactionController
import it.corelab.studios.airbooks.model.interfaces.login.signin.ErrorDialogController
import it.corelab.studios.airbooks.view.anko.layout.login.pages.RecoverPasswordUI
import org.jetbrains.anko.AnkoContext
import org.jetbrains.anko.support.v4.ctx


class RecoverPasswordFragment : Fragment(), RecoverPasswordController, ErrorDialogController, TransactionController {


    //VIEW
    private lateinit var viewUI: View


    //FRAGMENT MANAGER
    private var fragmentManagerRecover: FragmentManager? = null


    //VIEW VARIABLES
    private lateinit var email: EditText
    private lateinit var recoverButton: Button
    private lateinit var backButton: Button
    private val isValidCredential: Boolean get() = isEmailValid(getString(email))


    //API
    private var mAPIService: APIService? = null


    //UI INITIALIZATION
    private lateinit var mainUI: RecoverPasswordUI

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mainUI = RecoverPasswordUI()
        viewUI = mainUI.createView(AnkoContext.create(ctx, this))

        mAPIService = ApiUtils.getAPIService()

        initViews()

        setListeners()

        return viewUI
    }

    //==============================================================================================================================================================
    //==============================================================================================================================================================


    //RECOVER PASSWORD CONTROLLER
    override fun initViews() {

        fragmentManagerRecover = activity!!.supportFragmentManager

        email = mainUI.email
        recoverButton = mainUI.recoverButton
        backButton = mainUI.backButton
    }

    override fun setListeners() {

        backButton.setOnClickListener {
            fragmentManager!!
                    .beginTransaction()
                    .setCustomAnimations(R.anim.right_enter_animation, R.anim.left_exit_animation)
                    .replace(R.id.FRAME_CONTAINER_LOGIN_ACTIVITY, LoginFragment(), "Login_fragment")
                    .commit()
        }

        recoverButton.setOnClickListener{

            verifyCredential()

            if (isValidCredential) {

                Log.i("Credentials", "OK")
                sendPost(email.text.toString(), "http://airbooks.altervista.org/API/v2/recover.php?email=" + email.text.toString(), Locale.getDefault().language, "android")

            }

        }
    }

    override fun isEmailValid(email: CharSequence): Boolean {
        return android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }

    override fun getString(textInputEditText: EditText?): String {
        return textInputEditText!!.text.toString()
    }

    override fun verifyCredential() {
        if (!isEmailValid(getString(email))) {
            email.error = "you've to insert an email"
        }
    }

    override fun sendPost(emailText: String, url: String, lang: String, os: String) {


        val pDialog = SweetAlertDialog(activity!!, SweetAlertDialog.PROGRESS_TYPE)
        showProgressDialog(pDialog)

        mAPIService!!.savePost(emailText, url, lang, os).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<PostRecoverResponse> {

            override fun onSubscribe(d: Disposable) {

            }

            override fun onNext(postRecoverResponse: PostRecoverResponse) {
                pDialog.dismiss()
                if (postRecoverResponse.error != null) {
                    Log.i(TAG, "post submitted to API. " + postRecoverResponse.error.code!!.toString())
                    showErrorDialog()
                } else {
                    createDialog()
                }
            }

            override fun onError(e: Throwable) {

            }

            override fun onComplete() {

            }
        })
    }

    //==============================================================================================================================================================
    //==============================================================================================================================================================


    //ERROR DIALOG CONTROLLER
    override fun showErrorDialog() {
        SweetAlertDialog(activity!!, SweetAlertDialog.ERROR_TYPE)
                .setTitleText("Errore")
                .setContentText("Mail non esistente")
                .show()
    }

    override fun showProgressDialog(dialog: SweetAlertDialog?) {
        dialog!!.progressHelper.barColor = Color.parseColor("#4990e2")
        dialog.titleText = "Loading"
        dialog.setCancelable(false)
        dialog.show()
    }

    //==============================================================================================================================================================
    //==============================================================================================================================================================



    //TRANSACTION CONTROLLER
    override fun createDialog() {
        SweetAlertDialog(activity!!, SweetAlertDialog.WARNING_TYPE)
                .setTitleText("Recupero password")
                .setContentText("È stata inviata una mail per il recupero password al tuo indirizzo di posta elettronica")
                .setConfirmText("Ho capito")
                .setConfirmClickListener { sweetAlertDialog ->
                    fragmentManager!!
                            .beginTransaction()
                            .setCustomAnimations(R.anim.right_enter_animation, R.anim.left_exit_animation)
                            .replace(R.id.FRAME_CONTAINER_LOGIN_ACTIVITY, LoginFragment(), "Login_fragment")
                            .commit()

                    sweetAlertDialog.dismiss()
                }
                .show()
    }

    //==============================================================================================================================================================
    //==============================================================================================================================================================


}
